### PR TITLE
Add the option to resize an image only if the original image size is big...

### PR DESCRIPTION
**Pull request by UglyTroLL**
...ger than the target size.
#772

### CODE REVIEW

File: picasso/src/main/java/com/squareup/picasso/Request.java [hunk 1]
```diff
@@ -350,6 +354,21 @@ public Builder clearCenterInside() {
       centerInside = false;
       return this;
     }
+    
+    /**
+     * Only resize an image if the original image size is bigger than the target size 
+     * specified by {@link #resize(int, int)}.
+     */
+    public Builder onlyScaleDown() {
+        onlyScaleDown = true;
```
Comment (JakeWharton) [inline picasso/src/main/java/com/squareup/picasso/Request.java]: This needs to throw an exception is `resize` hasn't been called.

File: picasso/src/main/java/com/squareup/picasso/Request.java [hunk 2]
```diff
@@ -350,6 +354,21 @@ public Builder clearCenterInside() {
...
+        return this;
```
Comment (JakeWharton) [inline picasso/src/main/java/com/squareup/picasso/Request.java]: Also two-space indent

File: picasso/src/main/java/com/squareup/picasso/BitmapHunter.java [hunk 3]
```diff
@@ -480,7 +487,10 @@ static Bitmap transformResult(Request data, Bitmap result, int exifRotation) {
             targetWidth != 0 ? targetWidth / (float) inWidth : targetHeight / (float) inHeight;
         float sy =
             targetHeight != 0 ? targetHeight / (float) inHeight : targetWidth / (float) inWidth;
-        matrix.preScale(sx, sy);
+        if (!onlyScaleDown || (onlyScaleDown
+                && (inWidth > targetWidth || inHeight > targetHeight))) {
```
Comment (JakeWharton) [inline picasso/src/main/java/com/squareup/picasso/BitmapHunter.java]: Continuations should be indented with 4 spaces

File: picasso/src/test/java/com/squareup/picasso/BitmapHunterTest.java [hunk 4]
```diff
@@ -637,6 +637,30 @@
...
+    ShadowMatrix shadowMatrix = shadowOf(matrix);
+    
+    assertThat(shadowMatrix.getPreOperations()).containsOnly("scale 0.5 0.5");
+  }
+  
+  @Test public void onlyScaleDownOriginalSmaller() throws Exception {
+    Bitmap source = Bitmap.createBitmap(50, 50, ARGB_8888);
+    Request data = new Request.Builder(URI_1).resize(100, 100).onlyScaleDown().build();
+    Bitmap result = transformResult(data, source, 0);
+
```
Comment (JakeWharton) [inline picasso/src/test/java/com/squareup/picasso/BitmapHunterTest.java]: I think you want an `assertThat(result).isSameAs(source);` here as well

Comment (UglyTroLL) [inline picasso/src/test/java/com/squareup/picasso/BitmapHunterTest.java]: Added

File: picasso/src/main/java/com/squareup/picasso/BitmapHunter.java (hunk 3 above)
Comment (UglyTroLL) [inline picasso/src/main/java/com/squareup/picasso/BitmapHunter.java]: Fixed

File: picasso/src/main/java/com/squareup/picasso/Request.java (hunk 2 above)
Comment (UglyTroLL) [inline picasso/src/main/java/com/squareup/picasso/Request.java]: Fixed

File: picasso/src/main/java/com/squareup/picasso/Request.java (hunk 1 above)
Comment (UglyTroLL) [inline picasso/src/main/java/com/squareup/picasso/Request.java]: Added

File: picasso/src/test/java/com/squareup/picasso/BitmapHunterTest.java [hunk 5]
```diff
@@ -637,6 +637,31 @@
...
     assertThat(shadowMatrix.getPreOperations()).containsOnly("scale 0.5 0.5");
   }
+  
+  @Test public void onlyScaleDownOriginalBigger() throws Exception {
+    Bitmap source = Bitmap.createBitmap(100, 100, ARGB_8888);
+    Request data = new Request.Builder(URI_1).resize(50, 50).onlyScaleDown().build();
+    Bitmap result = transformResult(data, source, 0);
+    
+    ShadowBitmap shadowBitmap = shadowOf(result);
+    assertThat(shadowBitmap.getCreatedFromBitmap()).isSameAs(source);
```
Comment (JakeWharton) [inline picasso/src/test/java/com/squareup/picasso/BitmapHunterTest.java]: Hmm. This isn't what I meant. `result` and `source` should be the same instance.

Comment (UglyTroLL) [inline picasso/src/test/java/com/squareup/picasso/BitmapHunterTest.java]: This is inside onlyScaleDownOriginalBigger(), result and source shouldnt be the same. You probably meant the code in line 659

Comment (JakeWharton) [inline picasso/src/test/java/com/squareup/picasso/BitmapHunterTest.java]: I did! Thanks haha.

Comment (UglyTroLL) [inline picasso/src/test/java/com/squareup/picasso/BitmapHunterTest.java]: The commits were squashed and looks like Travis is happy

Comment (JakeWharton) [inline picasso/src/test/java/com/squareup/picasso/BitmapHunterTest.java]: I left for the day. Will look when I get home.

**Comment by JakeWharton** [comment]
Can you [sign our CLA][1] so we can merge this?

 [1]: http://squ.re/sign-the-cla

**Comment by UglyTroLL** [comment]
Done, thanks

File: picasso/src/main/java/com/squareup/picasso/BitmapHunter.java [hunk 6]
```diff
@@ -480,7 +487,10 @@ static Bitmap transformResult(Request data, Bitmap result, int exifRotation) {
             targetWidth != 0 ? targetWidth / (float) inWidth : targetHeight / (float) inHeight;
         float sy =
             targetHeight != 0 ? targetHeight / (float) inHeight : targetWidth / (float) inWidth;
-        matrix.preScale(sx, sy);
+        if (!onlyScaleDown || (onlyScaleDown
+            && (inWidth > targetWidth || inHeight > targetHeight))) {
```
Comment (dnkoutso) [inline picasso/src/main/java/com/squareup/picasso/BitmapHunter.java]: can you extract this conditional into a method since its copied 3 times?

**Comment by dnkoutso** [comment]
Minor nit: LGTM overall.

File: picasso/src/main/java/com/squareup/picasso/BitmapHunter.java (hunk 6 above)
Comment (UglyTroLL) [inline picasso/src/main/java/com/squareup/picasso/BitmapHunter.java]: Done, will squash if it LGTY

**Comment by dnkoutso** [comment]
danke!

**Pull request merged**